dvc/remote/base.py

```diff
@@ -180,7 +186,7 @@ def _collect_dir(self, path_info):
                 dir_info.append(
```

Comment:
Just make `get_file_checksum` return a dict `{"checksum": "mychecksum"}` and then use `.update({self.PARAM_RELPATH: relpath})`. Otherwise, you are splitting the knowtion of what checksum type `get_file_checksum` has returned and then assembling it all together, which is confusing.

Reply:
Well, that is not my code. Why was it accepted before?

Reply:
Keep packing and unpacking checksum and checksum type as a dict is not efficient and is unnecessary.
Because generating new checksum will always use the preferred checksum type, which is always known here.

```
@property
def checksums(self):
    return self._checksums

@property
def checksum(self):
    return self._checksums[0]
```

Reply:
> Because generating new checksum will always use the preferred checksum type, which is always known here.

It is not true though. Check out `remote.changed_checksum`, which needs to get_checksum of a particular type. Also, in "Remote.save_info" you are packing it into dict anyway. I'm just suggesting to keep it all together and not dispursing checksum notion everywhere. Check out your save_info:
```
        return {
            checksum_type
            or self.checksum_type(): self.get_checksum(
                path_info, checksum_type
            )
        }
```
with changes that I suggest it would be
```
return self.get_checksum(path_info)
```
🙂

Reply:
> It is not true though. Check out remote.changed_checksum, which needs to get_checksum of a particular type. Also, in "Remote.save_info" you are packing it into dict anyway. I'm just suggesting to keep it all together and not dispursing checksum notion everywhere. Check out your save_info:

That is why I said generating. 
When you are passing a particular type to `get_checksum`, you knows what type of checksum it is returning, isn't it?
🙂 It is a legacy code, not my packing.

Reply:
You do know if you specify that explicitly. But you are getting your checksum type from Remote while you are in the Output() class, and then you are passing it back to Remote. Doesn't seem very optimal to me. You could return `{"checksum": checksum}` and let your remote decide by itself which default checksum type it wants to use.

Reply:
> You do know if you specify that explicitly. But you are getting your checksum type from Remote while you are in the Output class, and then you are passing it back to Remote. Doesn't seem very optimal to me. You could return {"checksum": checksum} and let your remote decide by itself which default checksum type it wants to use.

About `But you are getting your checksum type from Remote while you are in the Output class, and then you are passing it back to Remote. `
Not sure about where you are talking, but
https://github.com/iterative/dvc/blob/00cfb04fea046946c0604985b3336e35bf7afd56/dvc/output/base.py#L143-L148
shows the code was doing that since the beginning 🙂 
And legacy code did deal with multiple checksums, because remotes don't use same checksums (types)

Reply:
Here https://github.com/iterative/dvc/pull/1920/files#diff-9487de78a43eff53118660a866090b89R144 you are getting checksum type from remote and using it later.

The legacy code was mostly built around the assumption of 1 checksum per remote, hence PARAM_CHECKSUM. The point of the task that you are doing is to generalize it and do it properly.

Reply:
The code was doing that since the beginning .🙂
Getting the checksum type from remote. `self.info.get(self.remote.PARAM_CHECKSUM)`
So, that does not change the fact that the legacy code still did deal with multiple checksums.